README.md
# channel_tools
Manipulate color and alpha using new "Channel Tools" in the GIMP "Layer"
menu.

## Install
### Linux
```
if [ -d ~/.config/GIMP/2.10/plug-ins/channel_tools.py ]; then
    rm ~/.config/GIMP/2.10/plug-ins/channel_tools.py
fi
cp channel_tools.py ~/.config/GIMP/2.10/plug-ins/
if [ -d ~/.config/GIMP/2.10/plug-ins/channel_tools ]; then
    rm ~/.config/GIMP/2.10/plug-ins/channel_tools  # try symlink FIRST
    rm -Rf ~/.config/GIMP/2.10/plug-ins/channel_tools
fi
cp -R channel_tools ~/.config/GIMP/2.10/plug-ins/
#or
ln -s ~/git/gimp-plugin-channel-tools/channel_tools.py ~/.config/GIMP/2.10/plug-ins/
ln -s ~/git/gimp-plugin-channel-tools/channel_tools ~/.config/GIMP/2.10/plug-ins/
```

## Developer Notes

### Python-fu sites
- [API Documentation](https://www.gimp.org/docs/python/index.html)
- [Gimp Scripting: Python Fu, Automating Workflows, coding a complete plug-in](https://www.youtube.com/watch?v=uSt80abcmJs) by Jason Bates Sep 13, 2015
- [Python fu #6: Accepting user input](https://jacksonbates.wordpress.com/2015/09/14/python-fu-6-accepting-user-input/) by Jason Bates


### GIMP API
- Booleans (non-zero if true):
  - drawable.is_color
  - drawable.has_alpha
  - drawable.is_gray
  - drawable.is_indexed
  - drawable.visible
- Other
  - image.active_channel (assignable)
  - image.cmap (color map)
  - image.layers (list of layers)
  - image.selection (selection mask)
  - image.add_layer_mask(layer, mask)



changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).


## [git] - 2020-02-20
### Added
- diffimage.py compares two images (calls tinkerduck.py)
- ChannelTinkerProgressInterface and ChannelTinkerInterface are
  available and allow the module to contain more of the functions that
  would otherwise be dependent upon the backend (GIMP or PIL).

### Changed
- Rename plug-in to channel_tinker a.k.a. "Channel Tinker" (and module
  to channel_tinker)
- Make the channel_tinker module entirely duck-typed.
    - Move gimp-specific usages to plugin script.


## [git] - 2020-02-19
### Added
- Draw Centered Square.


tests.py
#!/usr/bin/env python3

from diffimage import run as run_cli

run_cli("test_diff_base.png", "test_diff_head.png", "tmp.png")


channel_tools.py
#!/usr/bin/env python
"""
For each pixel where the alpha is below the threshold, get a new color
using the nearest opaque pixel.
"""
from gimpfu import *  # by convention, import *

from channel_tools import extend
from channel_tools import draw_square_from_center
from channel_tools import draw_circle_from_center
from channel_tools import convert_depth


def ct_draw_centered_circle(image, drawable, radius, color, filled):
    image.disable_undo()
    w = pdb.gimp_image_width(image)
    h = pdb.gimp_image_height(image)
    x = None
    y = None

    if w % 2 == 1:
        x = float(w) / 2.0
    else:
        x = w / 2
    if h % 2 == 1:
        y = float(h) / 2.0
    else:
        y = h / 2
    expand_right = False  # TODO: implement this
    expand_down = False  # TODO: implement this
    post_msg = ""
    if (w % 2 == 0):
        expand_right = True
        post_msg = "horizontally"
    if (h % 2 == 0):
        expand_down = True
        if len(post_msg) > 0:
            post_msg += " or "
        post_msg += "vertically"

    if len(post_msg) > 0:
        msg = ("The image is an even number of pixels, so the current"
               " drawing function cannot draw exactly centered"
               " " + post_msg + ".")
        pdb.gimp_message(msg)

    draw_circle_from_center((x, y), radius, image=image,
                            drawable=drawable,
                            color=color, filled=filled)
    pdb.gimp_drawable_update(drawable, 0, 0, drawable.width,
                             drawable.height)
    pdb.gimp_displays_flush()
    image.enable_undo()

def ct_draw_centered_square(image, drawable, radius, color, filled):
    image.disable_undo()
    w = pdb.gimp_image_width(image)
    h = pdb.gimp_image_height(image)
    x = w // 2
    y = h // 2
    expand_right = False  # TODO: implement this
    expand_down = False  # TODO: implement this
    post_msg = ""
    if (w % 2 == 0):
        expand_right = True
        post_msg = "horizontally"
    if (h % 2 == 0):
        expand_down = True
        if len(post_msg) > 0:
            post_msg += " or "
        post_msg += "vertically"

    if len(post_msg) > 0:
        msg = ("The image is an even number of pixels, so the current"
               " drawing function cannot draw exactly centered"
               " " + post_msg + ".")
        pdb.gimp_message(msg)

    print("image.channels: {}".format(image.channels))
    print("image.base_type: {}".format(image.channels))

    draw_square_from_center((x, y), radius, image=image,
                            drawable=drawable, color=color,
                            filled=filled)
    pdb.gimp_drawable_update(drawable, 0, 0, drawable.width,
                             drawable.height)
    pdb.gimp_displays_flush()
    image.enable_undo()


def ct_remove_layer_halo(image, drawable, minimum, maximum, good_minimum,
                         make_opaque, enable_threshold, threshold):
    gimp.progress_init("This may take a while...")
    # print("options: {}".format((str(image), str(drawable), str(minimum),
    #                            str(maximum), str(make_opaque),
    #                            str(good_minimum))))
    extend(image=image, drawable=drawable, minimum=minimum,
           maximum=maximum, make_opaque=make_opaque,
           good_minimum=good_minimum, enable_threshold=enable_threshold,
           threshold=threshold)
    pdb.gimp_drawable_update(drawable, 0, 0, drawable.width,
                             drawable.height)
    pdb.gimp_displays_flush()
    # ^ update the image; still you must first do#
    # pdb.gimp_drawable_update(...)
    image.enable_undo()

max_tip = "Maximum to discard (254 unless less damaged)"
goot_min_tip = "get nearby >= this (usually max discard+1)."
m_o_tip = "Make the fixed parts opaque."
a_t_tip = "Apply the threshold below to the image."
t_tip = "Minimum alpha to set to 255"
# See "Python-fu sites" under "Developer Notes" in README.md
register(
    "python_fu_ct_remove_halo",
    "Remove Halo",  # short description
    "Remove alpha",  # long description
    "Jake Gustafson", "Jake Gustafson", "2020",
    "Remove Halo",  # menu item caption
    "RGBA",  # RGB* would mean with or without alpha.
    [
        (PF_IMAGE, "image", "Current image", None),
        (PF_DRAWABLE, "drawable", "Input layer", None),
        (PF_SPINNER, "minimum", "Minimum alpha to fix (normally 1)", 1,
         (0, 255, 1)),
        (PF_SPINNER, "maximum", max_tip, 254, (0, 255, 1)),
        (PF_SPINNER, "good_minimum", goot_min_tip, 255, (0, 255, 1)),
        (PF_TOGGLE, "make_opaque", m_o_tip, False),
        (PF_TOGGLE, "enable_threshold", a_t_tip, False),
        (PF_SPINNER, "threshold", t_tip, 128, (0, 255, 1))
    ],
    [], # results
    ct_remove_layer_halo,
    menu="<Image>/Layer/Channel Tools"
)

# register(
    # "python_fu_ct_centered_circle",
    # "Draw Centered Circle",
    # "Draw centered circle",
    # "Jake Gustafson", "Jake Gustafson", "2020",
    # "Draw Centered Circle",  # caption
    # "RGB*",  # RGB* would mean with or without alpha.
    # [
        # (PF_IMAGE, "image", "Current image", None),
        # (PF_DRAWABLE, "drawable", "Input layer", None),
        # (PF_INT,    "radius", "Radius", 15),
        # (PF_COLOR,  "color", "Color", (0, 0, 0)),
        # (PF_TOGGLE, "filled", "Filled", False),
    # ],
    # [], # results
    # ct_draw_centered_circle,
    # menu="<Image>/Layer/Channel Tools"
# )


register(
    "python_fu_ct_centered_square",
    "Draw Centered Square",
    "Draw centered square",
    "Jake Gustafson", "Jake Gustafson", "2020",
    "Draw Centered Square",  # caption
    "RGB*",  # RGB* would mean with or without alpha.
    [
        (PF_IMAGE, "image", "Current image", None),
        (PF_DRAWABLE, "drawable", "Input layer", None),
        (PF_INT,    "radius", "Radius", 15),
        (PF_COLOR,  "color", "Color", (0, 0, 0)),
        (PF_TOGGLE, "filled", "Filled", False),
    ],
    [], # results
    ct_draw_centered_square,
    menu="<Image>/Layer/Channel Tools"
)

main()


channel_tinker_pil.py
#!/usr/bin/env python
from channel_tinker import diff_images
from PIL import Image
from channel_tinker import error

def diff_images_by_path(base_path, head_path, diff_path=None):
    """
    Compare two images.

    Keyword arguments:
    diff_path -- If not None, then save a graphical representation of
    the difference: black is same, closer to white differs (if images
    are different sizes, red is deleted, green is added).
    """
    base = Image.open(base_path)
    head = Image.open(head_path)
    w = max(base.size[0], head.size[0])
    h = max(base.size[1], head.size[1])
    diff_size = w, h
    diff = None
    # draw = None
    nochange_color = (0, 0, 0, 255)
    if diff_path is not None:
        diff = Image.new('RGBA', diff_size, nochange_color)
        # error("* generated diff image in memory")
        # draw = ImageDraw.Draw(diff)
    error("Checking {} zone...".format(diff_size))
    result = diff_images(base, head, diff_size, diff=diff,
                         nochange_color=nochange_color)
    if diff_path is not None:
        diff.save(diff_path)
        error("* saved {}".format(diff_path))
    return result




channel_tinker_gimp.py
#!/usr/bin/env python
"""
For each pixel where the alpha is below the threshold, get a new color
using the nearest opaque pixel.
"""

import math
import sys
from itertools import chain
import time

from gimpfu import *  # by convention, import *

from channel_tinker import convert_depth
from channel_tinker import error
from channel_tinker import idist
from channel_tinker import find_opaque_pos
from channel_tinker import ChannelTinkerInterface
from channel_tinker import ChannelTinkerProgressInterface
from channel_tinker import draw_circle_from_center
from channel_tinker import extend
from channel_tinker import draw_square_from_center


class GimpCTPI(ChannelTinkerProgressInterface):
    def progress_update(self, factor):
        gimp.progress_update(factor)

    def set_status(self, msg):
        gimp.progress_init(msg)

    def show_message(self, msg):
        pdb.gimp_message(msg)

class GimpCTI(ChannelTinkerInterface):

    @property
    def size(self):
        return self._size

    def __init__(self, image, drawable=None):
        self.image = image
        if drawable is None:
            drawable = pdb.gimp_image_active_drawable(image)
        self.drawable = drawable
        w = pdb.gimp_image_width(self.image)
        h = pdb.gimp_image_height(self.image)
        self._size = (w, h)
        self._bands = None
        self._p_len = None # for caching--not exposed

    def getbands(self):
        if self._bands is not None:
            return self._bands
        pos = (0, 0)
        p_len, pixel = pdb.gimp_drawable_get_pixel(self.drawable,
                                                   pos[0], pos[1])
        # TODO: return something more accurate if necessary.
        if p_len == 1:
            # See <https://stackoverflow.com/questions/52962969/
            # number-of-channels-in-pil-pillow-image>
            self._bands = tuple('L')  # Luma
            # TODO: handle 'P' (See <https://www.geeksforgeeks.org/
            #     pyhton-pil-getbands-method/>) [sic]
        else:
            self._bands = tuple("RGBA??????????"[:p_len])
        self._p_len = len(self._bands)
        return self._bands

    def getpixel(self, pos):
        p_len, pixel = pdb.gimp_drawable_get_pixel(self.drawable,
                                                   pos[0], pos[1])
        return pixel

    def putpixel(self, pos, color):
        if self._p_len is None:
            count = len(self.getbands())  # generates _p_len
        pdb.gimp_drawable_set_pixel(self.drawable, pos[0], pos[1],
                                    self._p_len, color)


def ct_draw_centered_circle(image, drawable, radius, color, filled):
    image.disable_undo()
    w = pdb.gimp_image_width(image)
    h = pdb.gimp_image_height(image)
    x = None
    y = None

    if w % 2 == 1:
        x = float(w) / 2.0
    else:
        x = w / 2
    if h % 2 == 1:
        y = float(h) / 2.0
    else:
        y = h / 2
    expand_right = False  # TODO: implement this
    expand_down = False  # TODO: implement this
    post_msg = ""
    if (w % 2 == 0):
        expand_right = True
        post_msg = "horizontally"
    if (h % 2 == 0):
        expand_down = True
        if len(post_msg) > 0:
            post_msg += " or "
        post_msg += "vertically"

    if len(post_msg) > 0:
        msg = ("The image is an even number of pixels, so the current"
               " drawing function cannot draw exactly centered"
               " " + post_msg + ".")
        pdb.gimp_message(msg)
    # exists, x1, y1, x2, y2 = \
    #     pdb.gimp_selection_bounds(self.image)
    cti = GimpCTI(image, drawable=drawable)
    draw_circle_from_center(cti, (x, y), radius,
                            color=color, filled=filled)
    pdb.gimp_drawable_update(drawable, 0, 0, drawable.width,
                             drawable.height)
    pdb.gimp_displays_flush()
    image.enable_undo()

def ct_draw_centered_square(image, drawable, radius, color, filled):
    image.disable_undo()
    w = pdb.gimp_image_width(image)
    h = pdb.gimp_image_height(image)
    x = w // 2
    y = h // 2
    expand_right = False  # TODO: implement this
    expand_down = False  # TODO: implement this
    post_msg = ""
    if (w % 2 == 0):
        expand_right = True
        post_msg = "horizontally"
    if (h % 2 == 0):
        expand_down = True
        if len(post_msg) > 0:
            post_msg += " or "
        post_msg += "vertically"

    if len(post_msg) > 0:
        msg = ("The image is an even number of pixels, so the current"
               " drawing function cannot draw exactly centered"
               " " + post_msg + ".")
        pdb.gimp_message(msg)

    print("image.channels: {}".format(image.channels))
    print("image.base_type: {}".format(image.channels))
    cti = GimpCTI(image, drawable=drawable)
    draw_square_from_center(cti, (x, y), radius, color=color,
                            filled=filled)
    pdb.gimp_drawable_update(drawable, 0, 0, drawable.width,
                             drawable.height)
    pdb.gimp_displays_flush()
    image.enable_undo()


def ct_remove_layer_halo(image, drawable, minimum, maximum, good_minimum,
                         make_opaque, enable_threshold, threshold):
    image.disable_undo()
    gimp.progress_init("This may take a while...")
    # print("options: {}".format((str(image), str(drawable), str(minimum),
    #                            str(maximum), str(make_opaque),
    #                            str(good_minimum))))
    cti = GimpCTI(image, drawable=drawable)
    ctpi = GimpCTPI()
    extend(cti, minimum=minimum,
           maximum=maximum, make_opaque=make_opaque,
           good_minimum=good_minimum, enable_threshold=enable_threshold,
           threshold=threshold, ctpi=ctpi)
    # if image is None:
    #     image = gimp.image_list()[0]
    # if drawable is None:
    #     drawable = pdb.gimp_image_active_drawable(image)

    pdb.gimp_drawable_update(drawable, 0, 0, drawable.width,
                             drawable.height)
    pdb.gimp_displays_flush()
    # ^ update the image; still you must first do#
    # pdb.gimp_drawable_update(...)
    image.enable_undo()

max_tip = "Maximum to discard (254 unless less damaged)"
goot_min_tip = "get nearby >= this (usually max discard+1)."
m_o_tip = "Make the fixed parts opaque."
a_t_tip = "Apply the threshold below to the image."
t_tip = "Minimum alpha to set to 255"
# See "Python-fu sites" under "Developer Notes" in README.md
register(
    "python_fu_ct_remove_halo",
    "Remove Halo",  # short description
    "Remove alpha",  # long description
    "Jake Gustafson", "Jake Gustafson", "2020",
    "Remove Halo",  # menu item caption
    "RGBA",  # RGB* would mean with or without alpha.
    [
        (PF_IMAGE, "image", "Current image", None),
        (PF_DRAWABLE, "drawable", "Input layer", None),
        (PF_SPINNER, "minimum", "Minimum alpha to fix (normally 1)", 1,
         (0, 255, 1)),
        (PF_SPINNER, "maximum", max_tip, 254, (0, 255, 1)),
        (PF_SPINNER, "good_minimum", goot_min_tip, 255, (0, 255, 1)),
        (PF_TOGGLE, "make_opaque", m_o_tip, False),
        (PF_TOGGLE, "enable_threshold", a_t_tip, False),
        (PF_SPINNER, "threshold", t_tip, 128, (0, 255, 1))
    ],
    [], # results
    ct_remove_layer_halo,
    menu="<Image>/Colors/Channel Tinker"
)

# register(
    # "python_fu_ct_centered_circle",
    # "Draw Centered Circle",
    # "Draw centered circle",
    # "Jake Gustafson", "Jake Gustafson", "2020",
    # "Draw Centered Circle",  # caption
    # "RGB*",  # RGB* would mean with or without alpha.
    # [
        # (PF_IMAGE, "image", "Current image", None),
        # (PF_DRAWABLE, "drawable", "Input layer", None),
        # (PF_INT,    "radius", "Radius", 15),
        # (PF_COLOR,  "color", "Color", (0, 0, 0)),
        # (PF_TOGGLE, "filled", "Filled", False),
    # ],
    # [], # results
    # ct_draw_centered_circle,
    # menu="<Image>/Colors/Channel Tinker"
# )


register(
    "python_fu_ct_centered_square",
    "Draw Centered Square",
    "Draw centered square",
    "Jake Gustafson", "Jake Gustafson", "2020",
    "Draw Centered Square",  # caption
    "RGB*",  # RGB* would mean with or without alpha.
    [
        (PF_IMAGE, "image", "Current image", None),
        (PF_DRAWABLE, "drawable", "Input layer", None),
        (PF_INT,    "radius", "Radius", 15),
        (PF_COLOR,  "color", "Color", (0, 0, 0)),
        (PF_TOGGLE, "filled", "Filled", False),
    ],
    [], # results
    ct_draw_centered_square,
    menu="<Image>/Colors/Channel Tinker"
)

main()
